chapter-05/final/money/decimal.go

```diff
@@ -12,61 +13,81 @@ import (
 type Number struct {
 	// the integer part of the Number
 	integerPart int
-	// the decimal part of the Number
-	decimalPart int
 	// precision of the decimal part, as the exponent of a power of 10
 	precision int
 }
 
 const (
-	// errors that we might face when parsing numbers
-	errInvalidInteger = moneyError("unable to convert integer part")
+	// errInvalidInteger is raised when parsing string into integer.
+	errInvalidInteger = moneyError("unable to parse integer")
 	errInvalidDecimal = moneyError("unable to convert decimal part")
 )
 
+// matchFloat regexp matches with a float
+var matchFloat = regexp.MustCompile("^((0|[1-9]\\d*)\\.(\\d*)?|\\.\\d+)$")
+
 // parseNumber converts a string into its Number representation.
 // It assumes there is up to one decimal separator, and that the separator is '.' (full stop character).
 func parseNumber(value string) (Number, error) {
 	intPart, decPart, found := strings.Cut(value, ".")
 
-	i, err := strconv.Atoi(intPart)
+	// validate the integer part
+	_, err := strconv.Atoi(intPart)
 	if err != nil {
 		return Number{}, fmt.Errorf("%w: %s", errInvalidInteger, err.Error())
 	}
 
-	var d int
+	// store the integer
+	fullValue := intPart
+
 	if found {
-		d, err = strconv.Atoi(decPart)
+		// validate the decimal part
+		_, err := strconv.Atoi(decPart)
 		if err != nil {
 			return Number{}, fmt.Errorf("%w: %s", errInvalidDecimal, err.Error())
 		}
+
+		// store the decimal
+		fullValue += decPart
 	}
 
+	// store the precision
 	precision := len(decPart)
 
-	return Number{integerPart: i, decimalPart: d, precision: precision}, nil
+	// convert the full value into an integer
+	number, err := strconv.Atoi(fullValue)
+	if err != nil {
+		return Number{}, fmt.Errorf("%w: %s", errInvalidInteger, err.Error())
+	}
+
+	return Number{integerPart: number, precision: precision}, nil
 }
 
 // applyChangeRate returns a new Number representing n multiplied by the rate.
 // The precision is the same in and out.
 func (n Number) applyChangeRate(rate ChangeRate, toPrecision int) Number {
+	// compute the new amount
 	converted := n.float() * float64(rate)
+	// shift the period to handle an integer
+	inFloat := converted * math.Pow10(toPrecision)
 
-	floor := math.Floor(converted)
-	decimal := math.Round((converted - floor) * math.Pow10(toPrecision))
+	// TODO: I am not sure if we should round or trunc here?
```

Comment:
trunc.
It's dangerous to round  and also not intuitive.

If you are in charge of a bureau of change, and someone gives you X that converts to 42.576, you are not going to give 42.58, but instead you'll return 42.57. In the first case, you gave/lost 0.004 from your own money, while in the second case, the customer paid/lost the 0.006.

Think of it this way:
If I convert X to the new currency, and the output is 1.0057, I can either floor it to 1.00 or round it to 1.01.
If 10000 people do that change, I will either have given away 10000, or I will have given away 10100. In the first case, my company has "won" 57 (by round to 1.00), and in the second case, it has lost 43 (by rounding to 1.00).